packages/eslint-plugin/src/rules/strict-enums.ts

```diff
@@ -0,0 +1,804 @@
+      if (type.isUnion()) {
+        const subTypes = tsutils.unionTypeParts(type);
+        const enumKinds = subTypes.map(subType => getEnumKind(subType));
+        return new Set(enumKinds);
+      }
+
+      if (type.isIntersection()) {
+        throw new Error(
+          'The getEnumKinds function is not meant to be used on intersections.',
+        );
+      }
+
+      const enumKind = getEnumKind(type);
+      return new Set([enumKind]);
+    }
+
+    /**
+     * A thing can have 0 or more enum types. For example:
+     * - 123 --> []
+     * - {} --> []
+     * - Fruit.Apple --> [Fruit]
+     * - Fruit.Apple | Vegetable.Lettuce --> [Fruit, Vegetable]
+     * - Fruit.Apple | Vegetable.Lettuce | 123 --> [Fruit, Vegetable]
+     * - T extends Fruit --> [Fruit]
+     */
+    function getEnumTypes(type: ts.Type): Set<ts.Type> {
+      /**
+       * First, we get all the parts of the union. For non-union types, this
+       * will be an array with the type in it. For example:
+       * - Fruit --> [Fruit]
+       * - Fruit | Vegetable --> [Fruit, Vegetable]
+       */
+      const subTypes = tsutils.unionTypeParts(type);
+
+      /**
+       * Next, we must resolve generic types with constraints. For example:
+       * - Fruit --> Fruit
+       * - T extends Fruit --> Fruit
+       */
+      const subTypesConstraints = subTypes.map(subType => {
+        const constraint = subType.getConstraint();
+        return constraint === undefined ? subType : constraint;
+      });
+
+      const enumSubTypes = subTypesConstraints.filter(subType =>
+        isEnum(subType),
+      );
+      const baseEnumSubTypes = enumSubTypes.map(subType =>
+        getBaseEnumType(subType),
+      );
+      return new Set(baseEnumSubTypes);
+    }
+
+    /**
+     * Given a set A and set B, return a set that contains only elements that are in
+     * both sets.
+     */
+    function getIntersectingSet<T>(a: Set<T>, b: Set<T>): Set<T> {
+      const intersectingValues = [...a.values()].filter(value => b.has(value));
+      return new Set(intersectingValues);
+    }
+
+    /**
+     * From:
+     * https://stackoverflow.com/questions/13627308/add-st-nd-rd-and-th-ordinal-suffix-to-a-number
+     */
+    function getOrdinalSuffix(i: number): string {
+      const j = i % 10;
+      const k = i % 100;
+      if (j == 1 && k != 11) {
+        return `${i}st`;
+      }
+      if (j == 2 && k != 12) {
+        return `${i}nd`;
+      }
+      if (j == 3 && k != 13) {
+        return `${i}rd`;
+      }
+      return `${i}th`;
+    }
+
+    function getTypeFromNode(node: TSESTree.Node): ts.Type {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      return getTypeFromTSNode(tsNode);
+    }
+
+    function getTypeFromTSNode(tsNode: ts.Node): ts.Type {
+      return typeChecker.getTypeAtLocation(tsNode);
+    }
+
+    function getTypeName(type: ts.Type): string {
+      return util.getTypeName(typeChecker, type);
+    }
```

Comment:
There's really no need for these two wrapper functions. They don't change anything.

```suggestion
```